gamemodes/terrortown/gamemode/server/sv_player_ext.lua

```diff
@@ -821,85 +826,192 @@ local function FindCorpsePosition(corpse)
 	return false
 end
 
+local spawnPositions = {}
+for i = 0, 360, 22.5 do
+	spawnPositions[#spawnPositions + 1] = Vector(mathCos(i), mathSin(i), 0)
+end
+spawnPositions[#spawnPositions + 1] = Vector(0, 0, 1)
+
+local function MakeSpawnPositionSafe(unsafePos)
+	local sizePlayer = Vector(32, 32, 72)
+	local startPos = unsafePos + Vector(0, 0, 0.5 * sizePlayer.z)
+
+	for i = 1, #spawnPositions do
+		local v = spawnPositions[i]
+		local pos = startPos + v * sizePlayer * 1.5
+
+		-- Make sure there is enough space around the player
+		local traceWalls = util.TraceHull({
+			start = pos,
+			endpos = pos,
+			mins = -0.5 * sizePlayer,
+			maxs = 0.5 * sizePlayer
+		})
+
+		if traceWalls.Hit then continue end
+
+		-- make sure the new position is on the ground
+		local traceGround = util.TraceLine({
+			start = pos,
+			endpos = pos - Vector(0, 0, sizePlayer.z),
+		})
+
+		if not traceGround.HitWorld then continue end
+
+		return pos - Vector(0, 0, 0.5 * sizePlayer.z)
+	end
+
+	return unsafePos
+end
+
+local function OnReviveFailed(ply, failMessage)
+	if isfunction(ply.OnReviveFailedCallback) then
+		ply.OnReviveFailedCallback(ply, failMessage)
+
+		ply.OnReviveFailedCallback = nil
+	else
+		LANG.Msg(ply, failMessage, nil, MSG_MSTACK_WARN)
+	end
+
+	net.Start("TTT2RevivalStopped")
+	net.Send(ply)
+end
+
 ---
 -- Revives a @{Player}
--- @param number delay the delay of the revive
--- @param function fn the @{function} that should be run if the @{Player} revives
--- @param function check an additional checking @{function}
--- @param boolean needcorpse whether the dead @{Player} CORPSE is needed
--- @param boolean force if the @{Player} revive is already forced. Useful if you have multiple reviving equipments
--- @param function onFail this @{function} is called if the revive fails
--- @realm server
-function plymeta:Revive(delay, fn, check, needcorpse, force, onFail)
+-- @param [default=3]number delay The delay of the revive
+-- @param [opt]function OnRevive The @{function} that should be run if the @{Player} revives
+-- @param [opt]function DoCheck An additional checking @{function}
+-- @param [default=false]boolean needsCorpse Whether the dead @{Player} @{CORPSE} is needed
+-- @param [default=false]boolean blockRound Stops the round from ending if this is set to true until the player is alive again
+-- @param [opt]function OnFail This @{function} is called if the revive fails
+-- @param [opt]Vector spawnPos The position where the player should be spawned, accounts for minor obstacles
+-- @realm server
+function plymeta:Revive(delay, OnRevive, DoCheck, needsCorpse, blockRound, OnFail, spawnPos)
 	local ply = self
 	local name = "TTT2RevivePlayer" .. ply:EntIndex()
 
-	if timer.Exists(name) or ply.reviving then return end
+	if timer.Exists(name) or ply:IsReviving() then return end
```

Comment:
It shouldn't be possible that a timer ```TTT2RevivePlayer...``` exist and the player is not reviving. What is the reason for checking that timer explicitely?

Reply:
It has been this way. Maybe it has a reason.

Reply:
Noone should create a timer named like that. If a timer with that name exists, it is completely fine to overwrite it with ```timer.Create()``` instead of canceling the revive.